### PR TITLE
Dynamically load model code from the Hub

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR allows a user to upload a modeling file along their model weights, and then lets the `Auto` classes load that model. The API is still a bit manual for now, it will be smoothed in follow-up PRs.

In terms of user of the model, the only thing to add is the flag `trust_remote_code=True` when using an auto class:
```
from transformers import AutoModel

model = AutoModel.from_pretrained("sgugger/test_dynamic_model", trust_remote_code=True)
```
This will load a `FakeModel` as defined in [this file](https://huggingface.co/sgugger/test_dynamic_model/blob/main/modeling.py).

For the person uploading the model there is a tiny bit more work. First save the code of the model (with all necessary imports) in a single python file, for instance `modeling.py`. In the rest of this PR description, I'll use MODELING_FILE to represent the name of that file without the suffix (so in my example of `modeling.py`, MODELING_FILE is "modeling"). That file needs to be in the repository you upload.

Second, add a filed `auto_map` to the configuration of the model you want to upload. This needs to be a dictionary with the name of auto classes as keys, and the full name of the corresponding modeling class as values. By full name, I mean MODELING_FILE.CLASS_NAME. For instance `modeling.FakeModel` for a class `FakeModel` defined in the `modeling.py` module. Here is an example:
```
config.auto_map = {"AutoModel": "modeling.FakeModel"}
```

This needs to be done before the model is saved, so that when doing the `model.save_pretrained` call, the config is saved with that field. Once this is done, push everything to the Hub, and you should be good!

### CODE REVIEW

**Comment by flozi00** [comment]
Forgive me for interjecting and asking, but does this allow custom modules to be uploaded to modelhub and loaded and run with the official Transformers library ? just to be sure.

If that's the case, do you plan to add some code scanning ? I think on some cases where this would be an security vulnerability for remote code execution

**Comment by sgugger** [comment]
The PR requires an extra argument in the call to `from_pretrained`: `allow_custom_model=True` to execute the code from the hub. A user should not pass along that flag without having scanned the code of the repo to make they trust it.

**Comment by flozi00** [comment]
Great, thanks for clearing this

File: src/transformers/models/auto/auto_factory.py [hunk 1]
```diff
@@ -377,13 +390,26 @@ def from_config(cls, config, **kwargs):
...
+        trust_remote_code = kwargs.pop("trust_remote_code", False)
         kwargs["_from_auto"] = True
         if not isinstance(config, PretrainedConfig):
             config, kwargs = AutoConfig.from_pretrained(
                 pretrained_model_name_or_path, return_unused_kwargs=True, **kwargs
             )
-
-        if type(config) in cls._model_mapping.keys():
+        if hasattr(config, "auto_map") and cls.__name__ in config.auto_map:
+            if not trust_remote_code:
```
Comment (LysandreJik) [inline src/transformers/models/auto/auto_factory.py:400]: I wonder if it wouldn't be good practice to force a revision here too. Either through raising an error (best) or through logging a warning (ok): if a public repository becomes trendy, it wouldn't be surprising to see the user update the code and for it to, at best break systems running the code, at worst run arbitrary code through `torch.load`

**Review by LysandreJik — COMMENTED** [review]
I'm having a few failures so I'll wait until you have updated the description of the PR before finishing my review

File: src/transformers/models/auto/auto_factory.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/models/auto/auto_factory.py:400]: That would make the API harder to use, no?

**Review by LysandreJik — APPROVED** [review]
This is in very good shape! I tried to break it as hard as I could but couldn't manage to do so!

Following our IRL discussions, the following point stood out: if defining a model that has a significantly different architecture than the original model (for example using a `BertConfig` as the model configuration, and then defining a totally different model to BERT), then the model can still be loaded in traditional BERT architectures; this is quite an edge case here, but we should keep that in mind as new configurations are allowed -> we should then make sure that models error out when loaded in architectures that don't support them.

Fantastic that it works out of the box for private models too. This should enable a myriad of use-cases.

Looks good to me, eagerly awaiting the next update enabling configuration/tokenizer creation!

File: src/transformers/models/auto/auto_factory.py (hunk 1 above)
Comment (julien-c) [reply src/transformers/models/auto/auto_factory.py:400]: After discussing with @LysandreJik, we will need to remember to parse this `auto_map` attribute in the hub so we know which models are custom code models

File: src/transformers/models/auto/auto_factory.py [hunk 2]
```diff
@@ -377,13 +390,31 @@ def from_config(cls, config, **kwargs):
...
+                raise ValueError(
+                    f"Loading {pretrained_model_name_or_path} requires you to execute the modeling file in that repo "
+                    "on your local machine. Make sure you have read the code there to avoid malicious use, then set "
+                    "the option `trust_remote_code=True` to remove this error."
+                )
+            if kwargs.get("revision", None) is None:
+                logger.warn(
+                    "Explicitly passing a `revision` is encouraged when loading a model with custom code to avoid "
+                    "breaking changes if the model maintainer update their code."
```
Comment (LysandreJik) [inline src/transformers/models/auto/auto_factory.py:409]: That's what I had in mind but feel free to disregard.

```suggestion
                    "Explicitly passing a `revision` is encouraged when loading a model with custom code to ensure "
                    "no malicious code has been contributed in a newer revision."
```

**Review by LysandreJik — APPROVED** [review]
LGTM, feel free to merge!

**Comment by p-christ** [comment]
hi, quick question about this PR:  

It lets people upload a custom model to the hub great, but it seems that in order to call that model using the API it still requires the model to be using one of the provided Pipelines?  Is there any way of also allowing a custom pipeline when calling the custom models?

**Comment by sgugger** [comment]
Custom pipelines is something we will add support for, it's on the roadmap, but for now you can't use the custom model in the API indeed.

**Pull request merged**